### DISCUSSION TITLE
Building reflector for different kinds of resources (found by annotations)

### DISCUSSION THREAD

**Discussion by aellwein** [kube-rs/kube#1257] [Q&A] [answered]
Hello,

i want to build a custom reflector for different kinds of resources (secrets, configmaps maybe others),
the reflected resources must have some custom annotation. How is this supposed to look like using kube-rs?

Thanks in advance.

**Comment by clux** [comment]
There are some examples of how to plug different resources into reflectors in the example folder, see in particular:

- https://github.com/kube-rs/kube/blob/main/examples/crd_reflector.rs#L36-L40
- https://github.com/kube-rs/kube/blob/main/examples/pod_reflector.rs
- https://github.com/kube-rs/kube/blob/main/examples/node_reflector.rs
- https://github.com/kube-rs/kube/blob/main/examples/secret_reflector.rs

**Reply by aellwein** [reply]
@clux i've seen the examples already. I wonder if i would need two different reflector instances for different resources, because they are typed on ``Api<T>``. So how to integrate different reflectors in one loop, is actually the question.

**Reply by clux** [reply, marked as answer]
Ah, apologies. The Kubernetes api can only return one type at a time, and the `reflector` is made to ensure data availability in the store for a that single ongoing list/watch call (even if the connection is temporarily lost), so it is typed to that item because of this.

We haven't seen a need to allow putting more things in one `Store` because you can put multiple reflectors in a struct:

```rust
struct MultiReflector {
    podstore: Store<Pod>,
    deploystore: Store<Deployment>,
}
```

and implement a lookup on this struct that directs to the correct store based on a dynamic type / object reference. you would still have to pull on __all__ the associated watch streams, but you can simplify that by [flattening the streams](https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.flatten) or similar.

**Reply by aellwein** [reply]
Thanks! Makes sense to me.
